### PR TITLE
Feat(medusa): file base routing

### PR DESCRIPTION
**Pull request by adrien2p**
<strike>Experimental [temporary doc](https://github.com/adrien2p/archipelago/blob/main/README.md)</strike>

### CODE REVIEW

**Comment by pevey** [comment]
Is this meant to pave the way toward migrating to a different base, like fastify?

Otherwise, it seems like introducing custom syntax for configuring routes has more detriment than benefit.  It is more mental overhead, and more hurdle for new users.

**Comment by adrien2p** [comment]
@pevey in the long term we would like to be agnostic indeed, and we already discussed to make it that way when needed. But the first thing we would like to solve is the difficulties the users seams to face when creating custom end points. So the idea is to simplify the way you create end point and make it closer to what people know as well as documenting it. 

Also, on another note, both systems coexists with this pr.

Finally, it seams that a majority of the users are familiar with this type of routing, but not necessarily with express. It seams logical that this approach could lower the bar for them on the server side routing.

**Comment by olivermrbl** [comment]
Just want to steelman @adrien2p's arguments:

> Is this meant to pave the way toward migrating to a different base, like fastify?

Partly yes, but this is primarily an attempt to eliminate friction points related to Express-specific features and concepts when building custom endpoints.

The tool will be backward compatible, so existing users familiar with the current pattern can opt in to this if they find it offers a better developer experience.

> and more hurdle for new users
> 

We believe this tool will actually shorten the time it takes for new users to create their first custom endpoint. With this tool, users can write their endpoints just like they would write any other JavaScript code. 

What about this approach do you think will create hurdles for users?

**Comment by olivermrbl** [comment]
> Is this meant to pave the way toward migrating to a different base, like fastify?

Partly yes, but this is primarily an attempt to eliminate friction points related to Express-specific features and concepts when building custom endpoints.

The tool will be backward compatible, so existing users familiar with the current pattern can opt in to this if they find it offers a better developer experience.

> and more hurdle for new users
> 

We believe this tool will actually shorten the time it takes for new users to create their first custom endpoint. With this tool, users can write their endpoints just like they would write any other JavaScript code. 

What about this approach do you think will create hurdles for users?

**Comment by pevey** [comment]
Having the new configuration approach coexist with the old way for some time is a good way to introduce it.  When I first saw the PR, I was envisioning all current route customizations and plugins having to be updated, which would be very disruptive.

Thanks for the additional perspectives.

File: packages/medusa/src/loaders/helpers/routing/types.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  DELETE = "DELETE",
+  PATCH = "PATCH",
+  OPTIONS = "OPTIONS",
+  HEAD = "HEAD",
+  ALL = "ALL",
+}
+/* eslint-enable no-unused-vars */
+
+export type RouteConfig = {
+  method?: RouteVerbs | "get" | "GET"
```
Comment (adrien2p) [inline packages/medusa/src/loaders/helpers/routing/types.ts:15]: todo: rm lowecase get

File: packages/medusa/src/loaders/helpers/routing/types.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  handlers: ((...args: unknown[]) => Promise<unknown>)[]
+}
+
+export type GlobalMiddlewareRouteConfig = {
+  method?: RouteVerbs | "get" | "GET" // default 'get'
```
Comment (adrien2p) [inline packages/medusa/src/loaders/helpers/routing/types.ts:20]: todo: rm lowecase get

**Pull request closed without merging**